q2_quality_control/_utilities.py

```diff
@@ -0,0 +1,421 @@
+    # inconsistent about adding/removing spaces so let's just be safe)
+    replacements = {t: t.replace(" ", "") for t in df.columns}
+    return df.rename(columns=replacements, inplace=False)
+
+
+def _compute_per_level_accuracy(exp, obs, metadata, depth):
+    results = []
+    vectors = {}
+    for level in range(1, depth + 1):
+        vectors[level] = {'exp': [], 'obs': []}
+        # collapse taxonomy strings to level
+        exp_collapsed = _collapse_table(exp, level)
+        obs_collapsed = _collapse_table(obs, level)
+        # compute stats for each sample individually
+        for sample in obs_collapsed.index:
+            result = [sample, level]
+            # if metadata are passed, map exp sample ID to value in metadata
+            if metadata is not None:
+                exp_id = metadata[sample]
+            else:
+                exp_id = sample
+            # concatenate obs/exp observations to align features
+            joined_table = pd.concat(
+                [exp_collapsed.loc[exp_id],
+                 obs_collapsed.loc[sample]], axis=1).fillna(0)
+            # split joined table apart again for computing stats
+            exp_vector = joined_table.iloc[:, 0]
+            obs_vector = joined_table.iloc[:, 1]
+            exp_features = exp_vector[exp_vector != 0]
+            obs_features = obs_vector[obs_vector != 0]
+            # Count observed taxa
+            observed_feature_count = len(obs_features)
+            observed_feature_ratio = (
+                observed_feature_count / len(exp_features))
+            result.extend([observed_feature_count, observed_feature_ratio])
+            # compute TAR/TDR
+            result.extend(compute_taxon_accuracy(exp_features, obs_features))
+            # compute linear least-squares regression results
+            result.extend(linregress(exp_vector, obs_vector))
+            results.append(result)
+            # store vectors for constructing regplots
+            vectors[level]['exp'].extend(exp_vector)
+            vectors[level]['obs'].extend(obs_vector)
+    results = pd.DataFrame(
+        results, columns=['sample', 'level', 'Observed Taxa',
+                          'Observed / Expected Taxa', 'TAR', 'TDR', 'Slope',
+                          'Intercept', 'r-value', 'P value', 'Std Err'])
+    results['r-squared'] = results['r-value']**2
+    return results, vectors
+
+
+# ported and modified from tax-credit with permission of nbokulich
+def compute_taxon_accuracy(exp, obs):
+    actual_obs_ids = set(exp.index)
+    expected_obs_ids = set(obs.index)
+
+    tp = len(actual_obs_ids & expected_obs_ids)
+    fp = len(actual_obs_ids - expected_obs_ids)
+    fn = len(expected_obs_ids - actual_obs_ids)
+
+    if tp > 0:
+        p = tp / (tp + fp)
+        r = tp / (tp + fn)
+    else:
+        p, r = 0, 0
+
+    return p, r
+
+
+def _find_nearest_common_lineage(feature, exp_features):
+    feature_depth = len(feature.split(';'))
```

Comment:
Do we need any whitespace trimming, as well?

Reply:
That is a good point — I am not sure, but I have added it into the `str.rstrip` below just in case. Do you think we should be stripping whitespace from the list of `expected_features` as well?

Reply:
I think `expected_features` are fine, because they are being collapsed by `q2-taxa`, which performs its [own trimming](https://github.com/qiime2/q2-taxa/blob/master/q2_taxa/_util.py#L18).